File: kotlin/src/main/kotlin/tasks/miniaturesUpdate/FurtrackMiniatureUrlResolver.kt
package tasks.miniaturesUpdate

import io.ktor.http.*
import web.client.CookieEagerHttpClient
import web.client.FastHttpClient
import web.client.GentleHttpClient
import web.client.HttpClientInterface
import web.url.FreeUrl
import web.url.Url

class FurtrackMiniatureUrlResolver(
    httpClient: HttpClientInterface? = null,
) : MiniatureUrlResolver {
    private val httpClient = httpClient ?: CookieEagerHttpClient(GentleHttpClient(FastHttpClient()))
    private val regex: Regex = Regex("^https://www.furtrack.com/p/(?<pictureId>\\d+)\$")

    override fun supports(url: String) = regex.containsMatchIn(url)

    override fun getMiniatureUrl(url: Url): String {
        val pictureId = regex.find(url.getUrl())?.groups?.get("pictureId")?.value
            ?: throw MiniatureUrlResolverException("Failed matching pictureId in the photo URL")

        val pictureUri = "https://orca2.furtrack.com/thumb/$pictureId.jpg"

        val response = httpClient.fetch(FreeUrl(pictureUri), HttpMethod.Head)

        if (200 != response.metadata.httpCode) {
            throw MiniatureUrlResolverException("Non-200 HTTP response code")
        }

        return pictureUri
    }
}

File: src/Photos/MiniatureUrlResolver/FurtrackMiniatureUrlResolver.php
<?php

declare(strict_types=1);

namespace App\Photos\MiniatureUrlResolver;

use App\Photos\MiniaturesUpdateException;
use App\Utils\Regexp\Pattern;
use App\Utils\Web\HttpClient\GentleHttpClient;
use App\Utils\Web\HttpClient\HttpClientInterface;
use App\Utils\Web\Url\FreeUrl;
use App\Utils\Web\Url\Url;
use Override;
use Symfony\Component\DependencyInjection\Attribute\Autowire;

class FurtrackMiniatureUrlResolver implements MiniatureUrlResolver
{
    private readonly Pattern $pattern;

    public function __construct(
        #[Autowire(service: GentleHttpClient::class)]
        private readonly HttpClientInterface $httpClient,
    ) {
        $this->pattern = new Pattern('^https://www\.furtrack\.com/p/(?<pictureId>\d+)$');
    }

    #[Override]
    public function supports(string $url): bool
    {
        return $this->pattern->isMatch($url);
    }

    #[Override]
    public function getMiniatureUrl(Url $url): string
    {
        $pictureId = $this->pattern->strictMatch($url->getUrl())->matches['pictureId'];
        $miniatureUrl = "https://orca2.furtrack.com/thumb/$pictureId.jpg";

        $response = $this->httpClient->fetch(new FreeUrl($miniatureUrl, $url->getCreatorId()), 'HEAD');

        if (200 !== $response->metadata->httpCode) {
            throw new MiniaturesUpdateException('Non-200 HTTP response code.');
        }

        return $miniatureUrl;
    }
}
